src/main/src/main/java/org/geoserver/security/decorators/ClipIntersectsFeatureIterator.java

```diff
@@ -0,0 +1,184 @@
+            boolean preserveZ) {
+        super(delegate, clip, schema, preserveZ);
+        this.intersects = intersects;
+    }
+
+    @Override
+    public boolean hasNext() {
+
+        while (next == null && delegate.hasNext()) {
+            // try building the clipped feature out of the original feature, if the
+            // default geometry is clipped out, skip it
+            SimpleFeature f = delegate.next();
+
+            boolean doTheClip = intersects == null ? true : false;
+
+            Map<Name, Geometry> intersectedGeometries = null;
+            if (intersects != null) {
+                Map<Name, Geometry> geometryAttributes = extractGeometryAttributes(f);
+                intersectedGeometries =
+                        getIntersectingGeometries(geometryAttributes, f.getFeatureType());
+                // if there is at least one geometryCollection or not all the geometry
+                // attributes were intersected performs also the clip
+                if (intersectedGeometries != null)
+                    doTheClip =
+                            intersectedGeometries
+                                            .values()
+                                            .stream()
+                                            .anyMatch(g -> g instanceof GeometryCollection)
+                                    || geometryAttributes.size() > intersectedGeometries.size();
+            }
+
+            boolean clippedOut = false;
+            if (doTheClip) clippedOut = prepareBuilderForNextFeature(f);
+
+            if (!clippedOut) {
+                // build the next feature
+                next = fb.buildFeature(f.getID());
+                unionWithIntersected(intersectedGeometries);
+
+            } else if (intersectedGeometries != null && !intersectedGeometries.isEmpty()) {
+                next = fb.buildFeature(f.getID());
+                for (Name name : intersectedGeometries.keySet()) {
+                    next.setAttribute(name, intersectedGeometries.get(name));
+                }
+            }
+
+            fb.reset();
+        }
+
+        return next != null;
+    }
+
+    // union the clipped geometries with the intersected one
+    private void unionWithIntersected(Map<Name, Geometry> intersectedGeometries) {
+        for (Name name : intersectedGeometries.keySet()) {
+            Geometry intersected = intersectedGeometries.get(name);
+            if (intersected != null && !intersected.isEmpty())
+                next.setAttribute(name, ((Geometry) next.getAttribute(name)).union(intersected));
+        }
+    }
+
+    private Map<Name, Geometry> getIntersectingGeometries(
+            Map<Name, Geometry> geometryAttributes, SimpleFeatureType type) {
+        Map<Name, Geometry> intersectedGeometries = new HashMap<>();
+        for (Name name : geometryAttributes.keySet()) {
+            Geometry geom = geometryAttributes.get(name);
+            if (geom instanceof GeometryCollection) {
```

Comment:
This bit seems weird too. A multi-geometry is still cosidered "one", while here it's taken apart in smaller bits?
Say the geometry is Italy and all its big and small islands, if I undrestand correctly, only the bits of that large multipolygon intersecting the geometry filter would be returned? That's now how spatial databases work, it's surprising behavior to me.

Reply:
fixed